Fix verifying prerequisites for local golang docker image

gouyang opened the pull request:
The image is prefixing with "docker.io", fix the regex.

Signed-off-by: Guohua Ouyang gouyang@redhat.com

gouyang commented:
I signed it

thockin commented on build/common.sh:
```diff
@@ -272,7 +272,7 @@ function kube::build::docker_image_exists() {
   # We cannot just specify the IMAGE here as `docker images` doesn't behave as
   # expected.  See: https://github.com/docker/docker/issues/8048
-  "${DOCKER[@]}" images | grep -Eq "^${1}\s+${2}\s+"
+  "${DOCKER[@]}" images | grep -Eq "${1}\s+${2}\s+"
```
Can you show me a line that docker emits that doesn't match the anchored regex?  I am trying to understand what changed...

thockin replied:
Nevermind - you said it in your comment.  I am not seeing docker.io - what docker version is doing this?

gouyang replied:
```
$ rpm -q docker
docker-1.7.0-6.git74e7a7a.fc22.x86_64
```

thockin replied:
I do not see this behavior.  I'm wary of loosening the regex without understanding why..

```console
thockin@freakshow kubernetes master /$ docker version
Client version: 1.7.1
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 786b29d
OS/Arch (client): linux/amd64
Server version: 1.7.1
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 786b29d
OS/Arch (server): linux/amd64

thockin@freakshow kubernetes master /$ docker pull busybox
latest: Pulling from busybox

cf2616975b4a: Already exists 
6ce2e90b0bc7: Already exists 
8c2e06607696: Already exists 
Digest: sha256:38a203e1986cf79639cfb9b2e1d6e773de84002feea2d4eb006b52004ee8502d
Status: Image is up to date for busybox:latest

thockin@freakshow kubernetes master /$ docker images | grep busybox
busybox                                     latest              8c2e06607696        3 months ago        2.433 MB

thockin@freakshow kubernetes master /$ docker rmi busybox
Untagged: busybox:latest
Deleted: 8c2e06607696bd4afb3d03b687e361cc43cf8ec1a4a725bc96e39f05ba97dd55
Deleted: 6ce2e90b0bc7224de3db1f0d646fe8e2c4dd37f1793928287f6074bc451a57ea
Deleted: cf2616975b4a3cba083ca99bc3f0bf25f5f528c3c52be1596b30f60b0b1c37ff

thockin@freakshow kubernetes master /$ docker pull docker.io/busybox
latest: Pulling from busybox

cf2616975b4a: Pull complete 
6ce2e90b0bc7: Pull complete 
8c2e06607696: Already exists 
busybox:latest: The image you are pulling has been verified. Important: image verification is a tech preview feature and should not be relied on to provide security.

Digest: sha256:38a203e1986cf79639cfb9b2e1d6e773de84002feea2d4eb006b52004ee8502d
Status: Downloaded newer image for busybox:latest

thockin@freakshow kubernetes master /$ docker images | grep busybox
busybox                                     latest              8c2e06607696        3 months ago        2.433 MB
```


Maybe someone from Docker can explain why you see a `docker.io/` prefix on images but I don't?  @jfrazzelle for a consult..

gouyang replied:
I'm on fedora 22, and I upgrade the docker to 1.7.1 by enable the update-testing repo.
It always pull from docker.io
```
$ docker pull nginx
latest: Pulling from docker.io/nginx
aface2a79f55: Pull complete 
5dd2638d10a1: Pull complete 
97df1ddba09e: Pull complete 
56c99fa886e8: Pull complete 
cd27805ea89f: Pull complete 
5e95ac0e9a79: Pull complete 
9f36f81a1ccb: Pull complete 
4c1809b04591: Pull complete 
98c9ccd75644: Pull complete 
6886fb5a9b8d: Already exists 
902b87aaaec9: Already exists 
9a61b6b1315e: Already exists 
docker.io/nginx:latest: The image you are pulling has been verified. Important: image verification is a tech preview feature and should not be relied on to provide security.
Digest: sha256:a2b8bef3338643176198510fec71c18ceb7834e6f7c45b039ec6e963f43b4005
Status: Downloaded newer image for docker.io/nginx:latest

$ docker version
Client version: 1.7.1.fc22
Client API version: 1.19
Package Version (client): docker-1.7.1-1.gitb6416b7.fc22.x86_64
Go version (client): go1.4.2
Git commit (client): b6416b7/1.7.1
OS/Arch (client): linux/amd64
Server version: 1.7.1.fc22
Server API version: 1.19
Package Version (server): docker-1.7.1-1.gitb6416b7.fc22.x86_64
Go version (server): go1.4.2
Git commit (server): b6416b7/1.7.1
OS/Arch (server): linux/amd64

$ docker images  nginx
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
docker.io/nginx     latest              6886fb5a9b8d        6 days ago          132.8 MB
```
And, I think it's more safe to remove the "^" in the regex, it's compatible with with prefix or without prefix.

thockin replied:
Can you tighten the regex to something like `^([a-z0-9-.]+/)?${1}\s` ?

thockin commented:
Can you split the docs stuff out?  It seems unrelated to the main point of this PR

thockin commented:
needs rebase

gouyang committed:
build: Fix verifying for local golang docker image

It always report missing local golang docker image if the local golang image
is prefixed with "docker.io".

Fixes #11732

Signed-off-by: Guohua Ouyang <gouyang@redhat.com>

gouyang commented:
@thockin 
I updated the PR, please review again, thanks.

zmerlynn commented:
Ah, sorry, this was superseded by #10824, which has been sitting for a while waiting to merge.

zmerlynn commented:
Though using this PR to tighten the regex would work fine as well. Either way. I allowed the `\S` through.

thockin commented:
I can live with ^\S/

On Mon, Jul 27, 2015 at 9:23 AM, Zach Loafman <notifications@github.com>
wrote:

> Though using this PR to tighten the regex would work fine as well. Either
> way. I allowed the \S through.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11284#issuecomment-125262258>
> .
>

gouyang commented:
I'm fine with the fixing by #10824

thockin commented:
Thanks for the PR, anyway :)

On Tue, Jul 28, 2015 at 5:32 AM, Guohua Ouyang <notifications@github.com>
wrote:

> I'm fine with the fixing by #10824
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10824>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11284#issuecomment-125584229>
> .
>
